### PR TITLE
Added support for SCE and NLL losses

### PR DESCRIPTION
**Pull request by abhishek-gola**
This pull request adds the support for Negative Log-Likelihood loss and Softmax Cross-Entropy loss in new DNN engine.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by dkurt** [comment]
DNN is an inference only module. Why do we need Losses operations support?

File: modules/dnn/src/layers/nll_loss_layer.cpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        auto weightAt = [&](int c) -> float {
+            if (!hasWeight) return 1.f;
+            CV_DbgAssert(0 <= c && c < C);
+            return w_f.at<float>(c, 0);
+        };
+
+        Mat logp2D; tensorNCX_to_NSxC(logp, logp2D);
+        Mat idx32;
+        if (label.type() == CV_32S) idx32 = label;
+        else                        label.convertTo(idx32, CV_32S);
```
Comment (asmorkalov) [inline modules/dnn/src/layers/nll_loss_layer.cpp]: Please reformat it as regular condition. It's not obvious for reader.

File: modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+                const float lp = logp2D.at<float>(r, y);
+                const float cw = hasWeight ? w.at<float>(y) : 1.f;
+
+                float loss = -(1.f - eps) * lp;
+                if (eps > 0.f) loss += -eps * meanLogRow[r];
+
+                per.at<float>(r) = loss * cw;
+                effW.at<float>(r) = cw;
+            }
+        }, 16);
```
Comment (asmorkalov) [inline modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp]: Formatting issue.

File: modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+            if (hasWeight) {
+                            for (int c = 0; c < C; ++c) {
+                const float wc = w.at<float>(c);
+                if (wc == 1.f) continue;
+                parallel_for_(Range(0, N*S), [&](const Range& rr){
+                    for (int r = rr.start; r < rr.end; ++r) lab2D.at<float>(r, c) *= wc;
+                }, 16);
+            }
+            }
```
Comment (asmorkalov) [inline modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp]: Formatting issue.

File: modules/dnn/src/onnx/onnx_importer2.cpp [hunk 4]
```diff
@@ -2564,6 +2580,9 @@ void ONNXImporter2::buildDispatchMap_ONNX_AI(int opset_version)
     dispatch["Tile"] = &ONNXImporter2::parseTile;
     dispatch["LayerNormalization"] = &ONNXImporter2::parseLayerNorm;
     dispatch["GroupNormalization"] = &ONNXImporter2::parseInstanceNormalization;
+    dispatch["NegativeLogLikelihoodLoss"] = &ONNXImporter2::parseNegativeLogLikelihoodLoss;
+    dispatch["SoftmaxCrossEntropyLoss"]   = &ONNXImporter2::parseSoftmaxCrossEntropyLoss;
+
```
Comment (asmorkalov) [inline modules/dnn/src/onnx/onnx_importer2.cpp:2642]: Extra empty line is not needed here.

File: modules/dnn/src/layers/nll_loss_layer.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                const float* nBase = srcData + n * sN;
+                for (int c = 0; c < C; ++c){
+                    dstRow[c] = nBase[c * sC + s];
+                }
+            }
+        }, nstripes);
+    }
+
+private:
+    String reduction;
```
Comment (vpisarev) [inline modules/dnn/src/layers/nll_loss_layer.cpp]: 1. attributes should be declared in the base `NegativeLogLikelihoodLossLayer`, not `NegativeLogLikelihoodLossLayerImpl`.
2. reduction should be enumeration, not string.

File: modules/dnn/src/layers/nll_loss_layer.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            }
+        }, nstripes);
+
+        if (reduction == "none")
+        {
+            std::vector<int> outShape; outShape.push_back(N);
+            for (int i = 2; i < logp.dims; ++i) {
+                outShape.push_back(logp.size[i]);
+            }
+            outv[0].create((int)outShape.size(), outShape.data(), CV_32F);
```
Comment (vpisarev) [inline modules/dnn/src/layers/nll_loss_layer.cpp]: when you call outv[0].create, it does not have any effect on the actual output array. You need to check  `out_arr.kind`, similar to implementation of many other such layers. Please, fix it.

In this particular case, since the engine pre-allocated tensors before calling forward, you should check the shape and type equality:

```
CV_Assert(outv[0].shape() == outShape && outv[0].type() == CV_32F);
```

File: modules/dnn/src/layers/nll_loss_layer.cpp [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            wflat.convertTo(w_f, CV_32F);
+        }
+
+        auto weightAt = [&](int c) -> float {
+            if (!hasWeight) return 1.f;
+            CV_DbgAssert(0 <= c && c < C);
+            return w_f.at<float>(c, 0);
+        };
+
+        Mat logp2D; tensorNCX_to_NSxC(logp, logp2D);
```
Comment (vpisarev) [inline modules/dnn/src/layers/nll_loss_layer.cpp]: try not to call intermediate conversion functions. Instead, create template function that will take indices of different types (CV_32S, CV_64S) and process them in one pass, not multiple passes.

File: modules/dnn/src/layers/nll_loss_layer.cpp [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        int sz3[3] = { N, C, S };
+        Mat src3 = src32.reshape(1, 3, sz3);
+
+        dst.create(N * S, C, CV_32F);
+
+        const size_t sN = src3.step1(0); // stride when n++
+        const size_t sC = src3.step1(1); // stride when c++
+
+        const float* srcData = src3.ptr<float>();
+        parallel_for_(Range(0, N * S), [&](const Range& r){
```
Comment (vpisarev) [inline modules/dnn/src/layers/nll_loss_layer.cpp]: why do we need shuffled version of inp[0]? Can the forward() method retrieve values immediately from inp[0]?

File: modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+                }
+            }
+            const float out = (reduction == "sum") ? static_cast<float>(num)
+                                                : static_cast<float>((den > 0.0) ? (num / den) : 0.0);
+            outv[0].create(1, 1, CV_32F);
+            outv[0].at<float>(0) = out;
+        }
+    }
+
+    static inline void tensorNCX_to_NSxC(const Mat& src, Mat& dst, int nstripes = 16)
```
Comment (vpisarev) [inline modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp]: is it the same function as in nll_loss_layer.cpp? If so, can we eliminate code duplication?

File: modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+            for (int r = rr.start; r < rr.end; ++r)
+            {
+                const float* x = logits2D.ptr<float>(r);
+                float m = x[0];
+                for (int c = 1; c < logits2D.cols; ++c) {
+                    m = std::max(m, x[c]);
+                }
+                double sum = 0.0;
+                for (int c = 0; c < logits2D.cols; ++c) {
+                    sum += std::exp(static_cast<double>(x[c]) - m);
```
Comment (vpisarev) [inline modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp]: this is softmax implementation, and we already have optimized softmax implementation. Can we share it instead of replication?

File: modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+                for (int c = 0; c < C; ++c) {
+                    nBase[c * sC + s] = srcRow[c];
+                }
+            }
+        }, nstripes);
+    }
+
+
+private:
+    String reduction;
```
Comment (vpisarev) [inline modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp]: 1. please, move attributes from the implementation to the base class
2. change the reduction type from string to enumeration. I think, the same enumeration can be used for nll_loss_layer.cpp.

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 12]
```diff
@@ -107,6 +107,14 @@ CV__DNN_INLINE_NS_BEGIN
         DNN_TARGET_CPU_FP16, // Only the ARM platform is supported. Low precision computing, accelerate model inference.
     };
 
+    // Shared reduction enum for DNN loss layers
+    enum LossReduction
+    {
+        DNN_LOSS_REDUCTION_NONE = 0,
```
Comment (vpisarev) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: for some enumerations we use DNN_ prefix, some some we don't. Can we normalize it?
Can you remove DNN_ prefix and move enumeration into all_layers.hpp?

File: modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+                const int s = r % S;
+                const float* rowBase = base + n * sN + s;
+
+                float maxv = rowBase[0];
+                for (int c = 1; c < C; ++c) maxv = std::max(maxv, rowBase[c * sC]);
+
+                double sumExp = 0.0;
+                double sumLogits = 0.0;
+                for (int c = 0; c < C; ++c) {
+                    const float v = rowBase[c * sC];
```
Comment (vpisarev) [inline modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp]: any vectorization of this code? scalar std::exp() is quite slow.

We have v_exp(), try to unroll the loop and use it.

File: modules/dnn/src/layers/nll_loss_layer.cpp [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+                                          const int ignoreIndex,
+                                          WeightGetter weightAt,
+                                          Mat& per,
+                                          Mat& effW,
+                                          Mat& validMask,
+                                          const int nstripes)
+    {
+        CV_Assert(idx1D.cols == 1);
+        parallel_for_(Range(0, idx1D.rows), [&](const Range& rr){
+            const float* base = logp32.ptr<float>();
```
Comment (vpisarev) [inline modules/dnn/src/layers/nll_loss_layer.cpp:189]: store values, accessed inside the innermost loop inside parallel_for_:

```
const T* idx1DData = idx1D.ptr<T>();
const float* wfData = w_f.ptr<float>();
uchar* validMaskData = validMask.ptr<uchar>();
float* errWData = errW.ptr<float>();
bool hasWeight = hasWeight_;
const int ignoreIndex = ignoreIndex_;
const in C = C_;
... // also S, sN, sC
```

also, move weightAt implementation inline into the parallel_for_ loop body.

File: modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+                const float* rowBase = base + n * sN_ + s;
+
+                float maxv = rowBase[0];
+                for (int c = 1; c < C_; ++c) maxv = std::max(maxv, rowBase[c * sC_]);
+
+                double sumExp = 0.0;
+                double sumLogits = 0.0;
+                for (int c = 0; c < C_; ++c) {
+                    const float v = rowBase[c * sC_];
+                    sumExp += std::exp(static_cast<double>(v - maxv));
```
Comment (vpisarev) [inline modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp]: 1. unroll the loop and use v_exp.
2. read data from tempbuf
3. whether double is necessary?

File: modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+            float* outlpPtr = outlpBase;
+            const size_t sN_out_ = sN_out;
+            const size_t sC_out_ = sC_out;
+            for (int r = rr.start; r < rr.end; ++r) {
+                const int n = r / S_;
+                const int s = r % S_;
+                const float* rowBase = base + n * sN_ + s;
+
+                float maxv = rowBase[0];
+                for (int c = 1; c < C_; ++c) maxv = std::max(maxv, rowBase[c * sC_]);
```
Comment (vpisarev) [inline modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp]: check softmax code. Do the same optimizations there:
1. allocate temporary vector of C_ elements (tempbuf) outside of `for (int r = rr.start; r < rr.end; ++r) {` loop
2. fetch data from rowBase to this buffer, compute max at once
3. compute sum of exponents, including the use of v_exp, compute sumLogits
4. store the results and optionally the output array.

File: modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+                    sumExp += std::exp(static_cast<double>(v - maxv));
+                    sumLogits += v;
+                }
+                const float lse = static_cast<float>(std::log(sumExp) + maxv);
+                rowLSE[r] = lse;
+                meanLR[r] = static_cast<float>(sumLogits / C_) - lse;
+
+                if (writeLP && outlpPtr) {
+                    float* dst = outlpPtr + n * sN_out_ + s;
+                    for (int c = 0; c < C_; ++c) dst[c * sC_out_] = rowBase[c * sC_] - lse;
```
Comment (vpisarev) [inline modules/dnn/src/layers/softmax_cross_entropy_loss_layer.cpp]: read data from tempbuf here, not rowBase

**Comment by asmorkalov** [comment]
@abhishek-gola Could you rebase and fix conflicts after https://github.com/opencv/opencv/pull/27882?
@vpisarev Do you have other remarks for the PR? I suppose it could be merged.

**Comment by asmorkalov** [comment]
@abhishek-gola I merged previous patch. Please go ahead.

**Comment by abhishek-gola** [comment]
@asmorkalov I have rebased and fixed the conflicts
